src/moonlink/src/row/column_array_builder.rs

```diff
@@ -95,9 +97,90 @@ impl ColumnArrayBuilder {
+            ),
             _ => panic!("data type: {data_type:?}"),
         }
     }
+
+    fn append_value_to_struct_field(
+        builder: &mut StructBuilder,
+        i: usize,
+        value: &RowValue,
+    ) -> Result<(), Error> {
+        if let Some(bool_builder) = builder.field_builder::<BooleanBuilder>(i) {
+            match value {
+                RowValue::Bool(v) => bool_builder.append_value(*v),
+                RowValue::Null => bool_builder.append_null(),
+                _ => unreachable!("Bool expected from well-typed input"),
+            }
+        } else if let Some(int32_builder) = builder.field_builder::<PrimitiveBuilder<Int32Type>>(i)
+        {
+            match value {
+                RowValue::Int32(v) => int32_builder.append_value(*v),
+                RowValue::Null => int32_builder.append_null(),
+                _ => unreachable!("Int32 expected from well-typed input"),
+            }
+        } else if let Some(int64_builder) = builder.field_builder::<PrimitiveBuilder<Int64Type>>(i)
+        {
+            match value {
+                RowValue::Int64(v) => int64_builder.append_value(*v),
+                RowValue::Null => int64_builder.append_null(),
+                _ => unreachable!("Int64 expected from well-typed input"),
+            }
+        } else if let Some(float32_builder) =
+            builder.field_builder::<PrimitiveBuilder<Float32Type>>(i)
+        {
+            match value {
+                RowValue::Float32(v) => float32_builder.append_value(*v),
+                RowValue::Null => float32_builder.append_null(),
+                _ => unreachable!("Float32 expected from well-typed input"),
+            }
+        } else if let Some(float64_builder) =
+            builder.field_builder::<PrimitiveBuilder<Float64Type>>(i)
+        {
+            match value {
+                RowValue::Float64(v) => float64_builder.append_value(*v),
+                RowValue::Null => float64_builder.append_null(),
+                _ => unreachable!("Float64 expected from well-typed input"),
+            }
+        } else if let Some(decimal_builder) =
+            builder.field_builder::<PrimitiveBuilder<Decimal128Type>>(i)
+        {
+            match value {
+                RowValue::Decimal(v) => decimal_builder.append_value(*v),
+                RowValue::Null => decimal_builder.append_null(),
+                _ => unreachable!("Decimal128 expected from well-typed input"),
+            }
+        } else if let Some(string_builder) = builder.field_builder::<StringBuilder>(i) {
+            match value {
+                RowValue::ByteArray(v) => {
+                    string_builder.append_value(unsafe { std::str::from_utf8_unchecked(v) })
+                }
+                RowValue::Null => string_builder.append_null(),
+                _ => unreachable!("String expected from well-typed input"),
+            }
+        } else if let Some(binary_builder) = builder.field_builder::<BinaryBuilder>(i) {
```

Comment:
another nit to mention is, we better reorder the field builder, so it's of the same order as column array builder:
https://github.com/Mooncake-Labs/moonlink/blob/23e38ada464dc60c294469af5f0ea42c2a85a4e7/src/moonlink/src/row/column_array_builder.rs#L31-L41
just easier to check if we miss anything.